File: SatellitesReactor/src/main/java/cz/radek/satellitesreactor/model/Cartesian.java
package cz.radek.satellitesreactor.model;

/**
 * Cartesian coordinates with basic operations
 * @author Tomsu Radek
 */
public class Cartesian {

	public static Cartesian CENTRE = new Cartesian( 0, 0, 0 );

	private final double x;
	private final double y;
	private final double z;

	public Cartesian( double x, double y, double z ) {
		this.x = x;
		this.y = y;
		this.z = z;
	}

	public double getX() {
		return x;
	}

	public double getY() {
		return y;
	}

	public double getZ() {
		return z;
	}

	/**
	 * Subtraction 
	 * @param c Cartesian
	 * @return Cartesian
	 */
	public Cartesian minus( Cartesian c ) {
		return new Cartesian( x - c.x, y - c.y, z - c.z );
	}

	/**
	 * Addition
	 * @param c Cartesian
	 * @return Cartesian
	 */
	public Cartesian plus( Cartesian c ) {
		return new Cartesian( x + c.x, y + c.y, z + c.z );
	}

	/**
	 * Scalar multiplication
	 * @param s Scalar
	 * @return Cartesian
	 */
	public Cartesian multiply( double s ) {
		return new Cartesian( x * s, y * s, z * s );
	}

	/**
	 * Vector product
	 * @param c Cartesian
	 * @return Cartesian
	 */
	public Cartesian vectorProduct( Cartesian c ) {
		return new Cartesian(
				y * c.z - c.y * z,
				x * c.z - c.x * z,
				x * c.y - c.x * y );
	}

	/**
	 * Dot product
	 * @param c Cartesian
	 * @return scalar
	 */
	public double dotProduct( Cartesian c ) {
		return x * c.x + y * c.y + z * c.z;
	}

	/**
	 * Distance to other point
	 * @param c Cartesian
	 * @return distance
	 */
	public double distanceTo( Cartesian c ) {
		return c.minus( this ).size();
	}

	/**
	 * Size of vector represented in Cartesian coordinates
	 * @return 
	 */
	public double size() {
		return Math.sqrt( Math.pow( x, 2 ) + Math.pow( y, 2 ) + Math.pow( z, 2 ) );
	}

	/**
	 * Squared size of vector represented in Cartesian coordinates
	 * @return 
	 */
	public double sizeSquared() {
		return Math.pow( x, 2 ) + Math.pow( y, 2 ) + Math.pow( z, 2 );
	}
}

File: SatellitesCpp/src/model/Cartesian.cpp
/*
 * Cartesian.cpp
 *
 */

#include "Cartesian.h"

#include <math.h>

double Cartesian::getX() const {
	return x_;
}

double Cartesian::getY() const {
	return y_;
}

double Cartesian::getZ() const {
	return z_;
}

Cartesian Cartesian::minus(const Cartesian& c) const {
	return Cartesian(x_ - c.x_, y_ - c.y_, z_ - c.z_);
}

Cartesian Cartesian::plus(const Cartesian& c) const {
	return Cartesian(x_ + c.x_, y_ + c.y_, z_ + c.z_);
}

Cartesian Cartesian::multiply(double s) const {
	return Cartesian(x_ * s, y_ * s, z_ * s);
}

Cartesian Cartesian::vectorProduct(const Cartesian& c) const {
	return Cartesian(y_ * c.z_ - c.y_ * z_, x_ * c.z_ - c.x_ * z_,
			x_ * c.y_ - c.x_ * y_);
}

double Cartesian::dotProduct(const Cartesian& c) const {
	return x_ * c.x_ + y_ * c.y_ + z_ * c.z_;
}

double Cartesian::distanceTo(const Cartesian& c) const {
	return c.minus(*this).size();
}

double Cartesian::size() const {
	return sqrt(pow(x_, 2) + pow(y_, 2) + pow(z_, 2));
}

double Cartesian::sizeSquared() const {
	return pow(x_, 2) + pow(y_, 2) + pow(z_, 2);
}


